### PR TITLE
Embedded GraphExp in JanusGraph POC Project

### PR DESCRIPTION
**Pull request by anki2189**
#https://github.com/strongbox/strongbox/issues/1609

### CODE REVIEW

**Comment by sbespalov** [comment]
@anki2189 I've tried to execute a query and it hangs with `Loading` message. Application logs:
```
15:25:49.000 29-11-2019 | DEBUG | remlin-server-exec-2 | o.a.t.gremlin.groovy.engine.GremlinExecutor        | Evaluating script - nodes = g.addV("ArtifactCoordinates").property("uuid", UUID.randomUUID().toString()).property("path","org/carlspring/test/dependency.jar").property("version","777").as("adc1").addV("ArtifactEntry").property("uuid", UUID.randomUUID().toString()).property("storageId", "storage0").property("repositoryId", "releases").as("ae1").addE("ArtifactDependency").to("adc1");edges = g.V(nodes).aggregate('node').outE().as('edge').inV().where(within('node')).select('edge').toList();[nodes,edges] - in thread [gremlin-server-exec-2]
15:25:49.148 29-11-2019 | DEBUG | mlin-server-worker-1 | o.a.t.gremlin.groovy.engine.GremlinExecutor        | Killing scheduled timeout on script evaluation - nodes = g.addV("ArtifactCoordinates").property("uuid", UUID.randomUUID().toString()).property("path","org/carlspring/test/dependency.jar").property("version","777").as("adc1").addV("ArtifactEntry").property("uuid", UUID.randomUUID().toString()).property("storageId", "storage0").property("repositoryId", "releases").as("ae1").addE("ArtifactDependency").to("adc1");edges = g.V(nodes).aggregate('node').outE().as('edge').inV().where(within('node')).select('edge').toList();[nodes,edges] - as the eval completed (possibly with exception).
```

Can you please check it?

**Comment by sbespalov** [comment]
also seems there is some layout issues

image

**Comment by sbespalov** [comment]
also can you please change the `<input type=text />` with something like `<textarea>` and change `Search` button label to `Execute`?

**Comment by sbespalov** [comment]
@anki2189 regarding the [layout](https://github.com/strongbox/strongbox-janusgraph-cassandra-poc/pull/15#issuecomment-559704382) I suggest to completely replace the Form Queries with just custom query text area.

**Comment by anki2189** [comment]
> g.addV("ArtifactCoordinates").property("uuid", UUID.randomUUID().toString()).property("path","org/carlspring/test/dependency.jar").property("version","777").as("adc1").addV("ArtifactEntry").property("uuid", UUID.randomUUID().toString()).property("storageId", "storage0").property("repositoryId", "releases").as("ae1").addE("ArtifactDependency").to("adc1")

Add .toList() at end of Query it will work.

Working Query : 
g.addV("ArtifactCoordinates").property("uuid", UUID.randomUUID().toString()).property("path","org/carlspring/test/dependency.jar").property("version","777").as("adc1").addV("ArtifactEntry").property("uuid", UUID.randomUUID().toString()).property("storageId", "storage0").property("repositoryId", "releases").as("ae1").addE("ArtifactDependency").to("adc1").toList()

**Comment by anki2189** [comment]
> @anki2189 I've tried to execute a query and it hangs with `Loading` message. Application logs:
> 
> ```
> 15:25:49.000 29-11-2019 | DEBUG | remlin-server-exec-2 | o.a.t.gremlin.groovy.engine.GremlinExecutor        | Evaluating script - nodes = g.addV("ArtifactCoordinates").property("uuid", UUID.randomUUID().toString()).property("path","org/carlspring/test/dependency.jar").property("version","777").as("adc1").addV("ArtifactEntry").property("uuid", UUID.randomUUID().toString()).property("storageId", "storage0").property("repositoryId", "releases").as("ae1").addE("ArtifactDependency").to("adc1");edges = g.V(nodes).aggregate('node').outE().as('edge').inV().where(within('node')).select('edge').toList();[nodes,edges] - in thread [gremlin-server-exec-2]
> 15:25:49.148 29-11-2019 | DEBUG | mlin-server-worker-1 | o.a.t.gremlin.groovy.engine.GremlinExecutor        | Killing scheduled timeout on script evaluation - nodes = g.addV("ArtifactCoordinates").property("uuid", UUID.randomUUID().toString()).property("path","org/carlspring/test/dependency.jar").property("version","777").as("adc1").addV("ArtifactEntry").property("uuid", UUID.randomUUID().toString()).property("storageId", "storage0").property("repositoryId", "releases").as("ae1").addE("ArtifactDependency").to("adc1");edges = g.V(nodes).aggregate('node').outE().as('edge').inV().where(within('node')).select('edge').toList();[nodes,edges] - as the eval completed (possibly with exception).
> ```
> 
> Can you please check it?
> 
> The original query was like below:
> 
> ```
> g.addV("ArtifactCoordinates").property("uuid", UUID.randomUUID().toString()).property("path","org/carlspring/test/dependency.jar").property("version","777").as("adc1").addV("ArtifactEntry").property("uuid", UUID.randomUUID().toString()).property("storageId", "storage0").property("repositoryId", "releases").as("ae1").addE("ArtifactDependency").to("adc1")
> ```

Only Search type Custom Queries were supported , I have added support for Add/update queries as well. Please try now.

File: src/main/resources/static/scripts/graphioGremlin.js [hunk 1]
```diff
@@ -159,13 +159,19 @@ var graphioGremlin = (function(){
...
+				    let gremlin_query_edges = "edges = " + traversal_source + ".V(nodes).aggregate('node').outE().as('edge').inV().where(within('node')).select('edge').toList();";
+	        	let gremlin_query = "nodes = " + custom_query + ";" + gremlin_query_edges + "[nodes,edges]";
+
+				    console.log(gremlin_query);
+	        	var message = "";
+	        	send_to_server(gremlin_query,'search',null,message);
+			} else {
+				    console.log(custom_query);
+	        	var message = "";
+	        	send_to_server(custom_query,'editGraph',null,message);
```
Comment (sbespalov) [inline src/main/resources/static/scripts/graphioGremlin.js]: @anki2189 what means 'editGraph' here?

**Comment by sbespalov** [comment]
@anki2189 you absolutely right, this was my fault and the query was incomplete. 
Anyway I just realized that we probably going a bit wrong way and all this changes need to be done within our [graphexp fork](https://github.com/strongbox/graphexp). It should be also packaged as `zip` and deployed into our artifacts repository. So can you please move all this changes there? 

@steve-todorov can help with the repository part, he knows better how we do it with `strongbox-web-ui`. 

So @anki2189 would you mind proceed like this?

After all this PR should just contain the `zip` artifact dependency and proper unpacking it into `static` resources.

**Comment by anki2189** [comment]
> @anki2189 you absolutely right, this was my fault and the query was incomplete.
> Anyway I just realized that we probably going a bit wrong way and all this changes need to be done within our [graphexp fork](https://github.com/strongbox/graphexp). It should be also packaged as `zip` and deployed into our artifacts repository. So can you please move all this changes there?
> 
> @steve-todorov can help with the repository part, he knows better how we do it with `strongbox-web-ui`.
> 
> So @anki2189 would you mind proceed like this?
> 
> After all this PR should just contain the `zip` artifact dependency and proper unpacking it into `static` resources.

Sure will proceed as suggested. Any other comments on code part ?
Layout overriding part is one thing which is left , will pick it later.

**Comment by sbespalov** [comment]
great. let's continue in the fork and it will be clear about new comments

**Comment by steve-todorov** [comment]
Closing this in favor of strongbox/strongbox#1622.

Thanks for your work and thoughts @sbespalov and @anki2189 and sorry for having to close this.

**Pull request closed without merging**